---
title: "Hippocampus RNAseq Nonlinear Modeling"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
#loading the packages we will be using for this analysis
library(zoo)
library(forecast)
library(reshape2)
library(parallel)
library(dplyr)
library(edgeR)
library(cluster)
library(factoextra)
library(tidyverse)
```

```{r}
# Load data - need residual counts matrix (removing technical effects) and metadata
library(dplyr)
library(edgeR)
# read/process relatedness matrix
relatedness <- read.csv('relatedness_matrix.csv')
rownames(relatedness) <- relatedness$X
relatedness$X <- NULL
colnames(relatedness) <- rownames(relatedness)

#Load in count matrix 
counts <- read.table('output.txt')
rownames(counts) <- counts$V1
counts$V1 <- NULL
colnames(counts) <- colnames(relatedness)
counts <- counts[2:34607,]
genes <- rownames(counts)
counts <- as.matrix(sapply(counts, as.numeric))
rownames(counts) <- genes

covariates <- read.csv('BrainSampleMetadata_RNASeq_22.csv')
covariates$Age_scaled <- scale(as.numeric(covariates$Age,scale=F, center=T))
covariates$SubjectID <- as.character(covariates$SubjectID)
covariates$Group <- as.factor(covariates$Group)
covariates$Lane <- as.factor(covariates$Lane)
covariates <- rename_with(covariates, tolower)
colnames(covariates)[1] <- "sampid"
head(covariates)
```

```{r}
#filtering and normalization
dge <- DGEList(counts = counts, group = covariates$age)
keep <- rowSums(cpm(counts, prior.count = 0.25, log = T) > 4) >= 15
table(keep)
dge <- dge[keep,,keep.lib.sizes=FALSE]
dge_TMM <- calcNormFactors(dge)

design <- model.matrix(~covariates$sex + covariates$age)
colnames(design) <- c('intercept','Sex','Age')
v <- voom(dge_TMM, design)
v <- subset(v$E)
v <- na.omit(v)

#remove 8 individuals from analysis that were outliers & had choroid plexus contamination
omit <- c("20267", "23298", "27796", "27381", "20616", "20698", "24699", "18621")
covariates <- filter(covariates, !(sampid %in% omit))
new_relatedness <- relatedness[c(covariates$sampid),c(covariates$sampid)]
v <- subset(v, select = c(covariates$sampid))
```

```{r}
save(v, covariates, file = "InputDataARIMA.RData")
```

### 3. Generating some PCAs and data figs

```{r}
#histogram summarizing sample info
# Create the histogram
samp <- ggplot(covariates, aes(x = age, fill = sex)) +
  geom_histogram(binwidth = 5, position = "stack") +
  labs(x = "Age", y = "Frequency") +
  scale_fill_manual(values = c("M" = "#98823c", "F" = "#9a5ea1")) +
  theme_minimal()

samp

ggsave(filename = "Sampleinfo_histogram_88.pdf", samp)
```

```{r}
design <- model.matrix(~covariates$chip + covariates$lane)
fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)


#counts_pca = prcomp(cor(resid_counts), center = T, scale = T)
counts_pca = prcomp(t(resid_counts))
summary(counts_pca)[["importance"]] %>%
  as.data.frame() %>%
  dplyr::select(c(PC1, PC2, PC3))
counts_pc <- as.data.frame(counts_pca$x) %>%
  rownames_to_column(var = "sampid")
counts_pc_meta <- left_join(counts_pc, covariates, by = "sampid")
summary(lm(PC1 ~ sex + age_scaled, counts_pc_meta))
```

```{r}
# #counts_pca = prcomp(cor(resid_counts), center = T, scale = T)
counts_pca = prcomp(t(v))
summary(counts_pca)[["importance"]] %>%
  as.data.frame() %>%
  dplyr::select(c(PC1, PC2, PC3))
counts_pc <- as.data.frame(counts_pca$x) %>%
  rownames_to_column(var = "sampid")
counts_pc_meta <- left_join(counts_pc, covariates, by = "sampid")
summary(lm(PC1 ~ sex + age_scaled, counts_pc_meta))

```

```{r}
p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=sex)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (23.14%)') + ylab('PC 2 (14.63%)') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18)) +
  scale_color_manual(values = c("M" = "#98823c", "F" = "#9a5ea1"))
plot(p)
ggsave(filename = "PCA_Sex_96.pdf", p)

p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=group)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (23.14%)') + ylab('PC 2 (14.63%)') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=age)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (23.14%)') + ylab('PC 2 (14.63%)') +
  scale_color_distiller(palette='GnBu') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)
ggsave(filename = "PCA_Age_96.pdf", p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=sex)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=age)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=group)) +
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

```

```{r}
p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=sex)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (24.87%)') + ylab('PC 2 (9.81%)') + 
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18)) +
  scale_color_manual(values = c("M" = "#98823c", "F" = "#9a5ea1"))
plot(p)
ggsave(filename = "PCA_Sex.pdf", p)

p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=group)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (24.87%)') + ylab('PC 2 (9.81%)') + 
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC1,y=PC2,color=age)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 1 (24.87%)') + ylab('PC 2 (9.81%)') +
  scale_color_distiller(palette='GnBu') +
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)
ggsave(filename = "PCA_Age.pdf", p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=sex)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') + 
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=age)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') + 
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)

p = ggplot(counts_pc_meta,aes(x=PC2,y=PC3,color=group)) + 
  geom_point(size=2.5) + theme_minimal() + xlab('PC 2 (9.81%)') + ylab('PC 3 (7.43%)') + 
  theme(axis.ticks=element_blank(),axis.text=element_blank(),axis.title = element_text(size=18), legend.text = element_text(size=18), legend.title = element_text(size=18))
plot(p)
```

```{r}
#Tests to see if PC1 is significantly associated with sex 
PC1_model <- lm(counts_pc$PC1 ~ covariates$age)
summary(lm(counts_pc$PC1 ~ covariates$age))
plot(PC1_model)
coef(PC1_model)
```
```{r}
#Making plots for genes for which there is choroid plexus contamination

library(ggplot2)
library(dplyr)

genes_up_fig <- as.data.frame(v['TTR',])
FOLR1 <- as.data.frame(v['FOLR1',])
PRLR <- as.data.frame(v['PRLR',])

genes_up_fig$FOLR1 <- FOLR1$`v["FOLR1", ]`
genes_up_fig$PRLR <- PRLR$`v["PRLR", ]`
colnames(genes_up_fig)[colnames(genes_up_fig) == 'v["TTR", ]'] <- 'TTR'

genes_up_fig$Age <- covariates$Age
genes_up_fig$Sex <- covariates$Sex

genes_up_fig = melt(genes_up_fig, id=c("Age", "Sex"))


# Create separate box plots for each gene
p <- ggplot(genes_up_fig, aes(x = variable, y = value, fill = variable)) +
  geom_boxplot() +
  labs(x = "Gene", y = "Gene Expression") +
  theme_minimal() +
  theme(legend.title = element_blank())
plot(p)
#ggsave(p, filename = "Choroid_Plexus_genes.pdf")
```
```{r}
#Making gene expression plots
genes_up_fig <- as.data.frame(fitted_nonzero['SPEN',])
rownames(genes_up_fig) <- gsub("X", "", rownames(genes_up_fig))


genes_up_fig$Age <- as.numeric(rownames(genes_up_fig))

p <- ggplot(genes_up_fig, aes(x = Age, y = `fitted_nonzero["SPEN", ]`)) +
  geom_point() +  # Scatter plot
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age", y = "Fitted expression counts") +
  scale_x_continuous(breaks = c(10, 20, 30))  # Set custom labels for the x-axis
plot(p)


genes_up_fig <- as.data.frame(fitted_nonzero['SIRT1',])
rownames(genes_up_fig) <- gsub("X", "", rownames(genes_up_fig))


genes_up_fig$Age <- as.numeric(rownames(genes_up_fig))

p <- ggplot(genes_up_fig, aes(x = Age, y = `fitted_nonzero["SIRT1", ]`)) +
  geom_point() +  # Scatter plot
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age", y = "Fitted expression counts") +
  scale_x_continuous(breaks = c(10, 20, 30))  # Set custom labels for the x-axis
plot(p)



genes_up_fig <- as.data.frame(fitted_nonzero['SIRT2',])
rownames(genes_up_fig) <- gsub("X", "", rownames(genes_up_fig))


genes_up_fig$Age <- as.numeric(rownames(genes_up_fig))

p <- ggplot(genes_up_fig, aes(x = Age, y = `fitted_nonzero["SIRT2", ]`)) +
  geom_point() +  # Scatter plot
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age", y = "Fitted expression counts") +
  scale_x_continuous(breaks = c(10, 20, 30))  # Set custom labels for the x-axis
plot(p)




```

```{r}
#Making input files for the ARIMAs models
#Model 1: regressing out effects of chips and sex 
design <- model.matrix(~covariates$chip + covariates$sex)
fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)

save(resid_counts, covariates, file = "InputDataARIMA_2.RData")
```

```{r}
# ARIMA fxn - from Marquez 2020
arima.selection <- function(adj.data=NULL,
                            metadata=NULL,
                            evalage=NULL,
                            use.boxcox=FALSE, # include selected.sex=NULL if you want to run ARIMA by sexes individually 
                            use.portmanteau=TRUE,
                            loess.span=0.75,
                            dataType=c("rna"),
                            n.cores=20) {
  require(zoo)
  require(forecast)
  require(reshape2)
  require(parallel)
  require(dplyr)
  
  age=with(metadata,setNames(age,sampid)) # pairs LID with age 
  sex=with(metadata,setNames(sex,sampid)) # pairs LID with sex (but this code does not separate GE by sex, original Marquez code does)
  
  cat("Building time series object\n")
  age.sexsel <- age[names(sex)]
  age.sexsel.sorted <- age.sexsel[order(age.sexsel)] # orders by age
  adj.data.sexsel <- adj.data[,names(sex)]
  
  selpeaks.sexsel <- t(adj.data.sexsel)
  age.counts <- as.data.frame(table(age.sexsel),stringsAsFactors = F)
  selpeaks.sexsel.unique <- do.call(rbind,lapply(setNames(as.list(seq_len(nrow(age.counts))),age.counts$age.sexsel),function(n) {
    if (age.counts[n,"Freq"]==1) {
      estage <- selpeaks.sexsel[age.sexsel==age.counts[n,"age.sexsel"],]
    } else {
      estage <- colMeans(selpeaks.sexsel[age.sexsel==age.counts[n,"age.sexsel"],])
    }
    return(estage)}))
  age.sexsel.unique <- as.numeric(age.counts$age.sexsel)
  
  # Create zoo/ts objects, average data for same ages, interpolate unsampled ages
  zoo.sexsel <- zoo(selpeaks.sexsel.unique,order.by=age.sexsel.unique,frequency = 1)
  ts.sexsel <- mclapply(zoo.sexsel,ts,mc.cores = n.cores)
  # plot.ts(ts.sexsel[[1]],type="b") # Visualization of a single TS
  if (use.boxcox) {
    cat("Calculating Box-Cox Lambda using the log-likelihood method\n")
    bclambda.sexsel <- sapply(ts.sexsel,BoxCox.lambda,method="loglik",upper=10)
  } else {
    bclambda.sexsel = NULL
  }
  gc()
  
  cat("Fitting auto-arima models on",n.cores,"CPU cores\n")
  
  # Builds an ARIMA model to test for trends in each time series, then selects those for which the trend is non-negligible (stationarity cannot be rejected AND no AR or MA terms are chosen)
  arima.sexsel <- mclapply(setNames(seq_len(length(ts.sexsel)),names(ts.sexsel)),function(n) auto.arima(ts.sexsel[[n]],seasonal = F,allowmean = T,allowdrift = T,lambda = bclambda.sexsel), mc.cores = n.cores)
  gc()
  if (use.portmanteau) {
    # portmanteau tests on RESIDUALS (Ljung-Box) - models with autocorrelated residuals are removed as well
    cat("Computing portmanteau tests on ARIMA model residuals using the Ljung-Box algorithm, on",n.cores,"CPU cores\n")
    portmanteau.sexsel <- unlist(mclapply(arima.sexsel,function(x) {
      df=sum(x$arma[1:2])
      pmin <- Box.test(x$residuals,lag=20,type="Ljung-Box",fitdf = df)$p.value
      return(pmin)
    },mc.cores = n.cores))
    cat("Proceeding to filter out",ifelse(dataType[[1]]=="atac","peaks","transcripts"),"with stationary models and autocorrelated residuals\n")
    arima.sexsel.select <- unlist(lapply(arima.sexsel,function(x) sum(x$arma[1:2])>0 & x$arma[6]>0)) & portmanteau.sexsel>0.05
  } else {
    cat("Proceeding to filter out",ifelse(dataType[[1]]=="atac","peaks","transcripts"),"with stationary models and autocorrelated residuals\n")
    arima.sexsel.select <- unlist(lapply(arima.sexsel,function(x) sum(x$arma[1:2])>0 & x$arma[6]>0))
    portmanteau.sexsel = NULL
  }
  gc()
  
  arima.sexsel.nonzero <- arima.sexsel[arima.sexsel.select]
  ts.sexsel.nonzero <- ts.sexsel[arima.sexsel.select]
  fitted.sexsel.nonzero <- t(as.data.frame(sapply(arima.sexsel.nonzero,`[[`,"fitted"),row.names = as.integer(attr(arima.sexsel.nonzero[[1]]$x,"index"))))
  cat("Out of",length(arima.sexsel),ifelse(dataType[[1]]=="atac","peaks,","transcripts,"),length(ts.sexsel.nonzero),"trendy and well-fitted",ifelse(dataType[[1]]=="atac","peaks","transcripts"),"were selected for further analysis\n")
  gc()
  
  ## Uses local fit (LOESS) on sex-union nonzero trends to predict chromatin values at each age in the span, for sex comparisons
  cat("Computing predicting",ifelse(dataType[[1]]=="atac","peaks","transcripts"),"using LOESS interpolation over ARIMA-fitted data on",n.cores,"cores\n")
  loess.sexsel.nonzero <- mclapply(arima.sexsel.nonzero,function(A) loess(as.vector(A$fitted)~attr(A$x,"index"),span = loess.span,family = "symmetric"),mc.cores = n.cores)
  gc()
  predicted.sexsel.nonzero <- t(as.data.frame(mclapply(loess.sexsel.nonzero,predict,evalage,mc.cores = n.cores),row.names = evalage))
  gc()
  
  output <- list(ts.complete=ts.sexsel,
                 arima.complete=arima.sexsel,
                 arima.select=arima.sexsel.select,
                 loess.nonzero=loess.sexsel.nonzero,
                 fitted.nonzero=fitted.sexsel.nonzero,
                 predicted.nonzero=predicted.sexsel.nonzero,
                 BoxCox_lambda=bclambda.sexsel,
                 portmanteau.pvalues=portmanteau.sexsel)
  
  cat("Done.\n")
  return(output)
}
```

```{r}
## Run with: 
aging_arima <- arima.selection(adj.data = v, metadata = covariates)
save(aging_arima, file = "aging_arima.RData")
```

```{r}
load("ARIMA_2/aging_sex_arima.RData")
```

```{r}
fitted_nonzero <- aging_arima$fitted.nonzero
predicted_nonzero <- aging_arima$predicted.nonzero

dim(fitted_nonzero)
dim(predicted_nonzero)
```

```{r}
# dist_matrix <- get_dist(predicted_nonzero, method = "spearman")
# head(dist_matrix)
# fviz_dist(dist_matrix, gradient = list(low = "#00AFBB", mid = "white", high = "#FC4E07"))
# 
# k5 <- kmeans(predicted_nonzero, centers = 5, nstart = 25)
# str(k5)
# 
# fviz_cluster(k5, data = predicted_nonzero)
```

```{r}
# fviz_nbclust(predicted_nonzero, kmeans, method = "wss")
# fviz_nbclust(predicted_nonzero, kmeans, method = "silhouette")
# gap_stat <- clusGap(predicted_nonzero, FUN = kmeans, nstart = 25,
#                     K.max = 10, B = 50)
# fviz_gap_stat(gap_stat)
```

```{r}
# hclust_matrix <- predicted_nonzero %>% 
#   # transpose the matrix so genes are as columns
#   t() %>% 
#   # apply scalling to each column of the matrix (genes)
#   scale() %>% 
#   # transpose back so genes are as rows again
#   t()
# 
# gene_dist <- dist(hclust_matrix)
# 
# gene_hclust <- hclust(gene_dist, method = "complete")
# 
# # The default `plot()` function can be used to produce a simple dendrogram
# plot(gene_hclust, labels = FALSE)
# abline(h = 10, col = "brown", lwd = 2) # add horizontal line to illustrate cutting dendrogram
```

```{r}
# gene_cluster <- cutree(gene_hclust, k = 5) %>% 
#   # turn the named vector into a tibble
#   enframe() %>% 
#   # rename some of the columns
#   rename(gene = name, cluster = value)
# 
# head(gene_cluster)
```

```{r}
# fitted_nonzero <- as.data.frame(fitted_nonzero)
# fitted_nonzero$cluster <- gene_cluster$cluster
# dim(fitted_nonzero)
# 
# cluster1_names <- subset(gene_cluster, gene_cluster$cluster == 1)
# cluster1 <- fitted_nonzero[fitted_nonzero[,88]==1,c(1:87)]
# rownames(cluster1) <- cluster1_names$gene
# cluster1 <- as.matrix(cluster1)
# longcluster1 <- melt(cluster1)
# longcluster1 <- na.omit(longcluster1)
# 
# heatmap(cluster1)
```

```{r}
# cluster1_names <- subset(gene_cluster, gene_cluster$cluster == 1)
# cluster1 <- fitted_nonzero[fitted_nonzero[,88]==1,c(1:87)]
# rownames(cluster1) <- cluster1_names$gene
# cluster1 <- as.matrix(cluster1)
# longcluster1 <- melt(cluster1)
# longcluster1 <- na.omit(longcluster1)
# 
# heatmap(cluster1, Colv = NA, Rowv = NA)
# 
# cluster2_names <- subset(gene_cluster, gene_cluster$cluster == 2)
# cluster2 <- fitted_nonzero[fitted_nonzero[,88]==2,c(1:87)]
# rownames(cluster2) <- cluster2_names$gene
# cluster2 <- as.matrix(cluster2)
# 
# heatmap(cluster2, Colv = NA, Rowv = NA)
# 
# cluster3_names <- subset(gene_cluster, gene_cluster$cluster == 3)
# cluster3 <- fitted_nonzero[fitted_nonzero[,88]==3,c(1:87)]
# rownames(cluster3) <- cluster3_names$gene
# cluster3 <- as.matrix(cluster3)
# 
# heatmap(cluster3, Colv = NA, Rowv = NA)
# 
# cluster4_names <- subset(gene_cluster, gene_cluster$cluster == 4)
# cluster4 <- fitted_nonzero[fitted_nonzero[,88]==4,c(1:87)]
# rownames(cluster4) <- cluster4_names$gene
# cluster4 <- as.matrix(cluster4)
# 
# heatmap(cluster4, Colv = NA, Rowv = NA)
# 
# cluster5_names <- subset(gene_cluster, gene_cluster$cluster == 5)
# cluster5 <- fitted_nonzero[fitted_nonzero[,88]==5,c(1:87)]
# rownames(cluster5) <- cluster5_names$gene
# cluster5 <- as.matrix(cluster5)
# 
# heatmap(cluster5, Colv = NA, Rowv = NA)
```

```{r}
library(heatmap3)
clus.distance.arima = as.dist(1 - cor(t(predicted_nonzero), use = "pa"))
h.clust.arima = hclust(clus.distance.arima,method='complete') # calc. clusters

gene_cluster <- cutree(h.clust.arima, k = 4) %>% 
  # turn the named vector into a tibble
  enframe() %>% 
  # rename some of the columns
  rename(gene = name, cluster = value)

head(gene_cluster)
```

```{r}
fitted_nonzero <- as.data.frame(fitted_nonzero)
fitted_nonzero$cluster <- gene_cluster$cluster
dim(fitted_nonzero)

cluster1_names <- subset(gene_cluster, gene_cluster$cluster == 1)
cluster1 <- fitted_nonzero[fitted_nonzero[,88]==1,c(1:87)]
rownames(cluster1) <- cluster1_names$gene
cluster1 <- as.matrix(cluster1)
longcluster1 <- melt(cluster1)
longcluster1 <- na.omit(longcluster1)

par(mar=c(1,1,1,1))
heatmap(cluster1, Colv = NA, Rowv = NA)

cluster2_names <- subset(gene_cluster, gene_cluster$cluster == 2)
cluster2 <- fitted_nonzero[fitted_nonzero[,88]==2,c(1:87)]
rownames(cluster2) <- cluster2_names$gene
cluster2 <- as.matrix(cluster2)

heatmap(cluster2, Colv = NA, Rowv = NA)

cluster3_names <- subset(gene_cluster, gene_cluster$cluster == 3)
cluster3 <- fitted_nonzero[fitted_nonzero[,88]==3,c(1:87)]
rownames(cluster3) <- cluster3_names$gene
cluster3 <- as.matrix(cluster3)

heatmap(cluster3, Colv = NA, Rowv = NA)

cluster4_names <- subset(gene_cluster, gene_cluster$cluster == 4)
cluster4 <- fitted_nonzero[fitted_nonzero[,88]==4,c(1:87)]
rownames(cluster4) <- cluster4_names$gene
cluster4 <- as.matrix(cluster4)

heatmap(cluster4, Colv = NA, Rowv = NA)


write_csv(cluster1_names, "cluster1all.csv")
write_csv(cluster2_names, "cluster2all.csv")
write_csv(cluster3_names, "cluster3all.csv")
write_csv(cluster4_names, "cluster4all.csv")
```

```{r}
overlapgenes <- merge(predicted_nonzero, NewAgeCovEffects_Genes, by="row.names")
dim(overlapgenes)
row.names(NewAgeCov_Upregulated) <- NewAgeCov_Upregulated$Gene
row.names(NewAgeCov_Downregulated) <- NewAgeCov_Downregulated$Gene
cluster1overlap <- merge(cluster1, NewAgeCov_Upregulated, by="row.names")
cluster2overlap <- merge(cluster2, NewAgeCov_Downregulated, by="row.names")
cluster3overlap <- merge(cluster3, NewAgeCovEffects_Genes, by="row.names")
cluster4overlap <- merge(cluster4, NewAgeCovEffects_Genes, by="row.names")
```

```{r}
#rerunning the ARIMA models with regressing out the effects of sex and the chips 
#### extract residuals from our model of interest
design <- model.matrix(~covariates$chip + covariates$sex)
fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)

save(resid_counts, covariates, file = "InputSexARIMA_2.RData")
```

```{r}
#Next we can load the results of this model in and begin to see the output

load("ARIMA_2/aging_sex_arima.RData")

Covfitted_nonzero <- aging_arima$fitted.nonzero
Covpredicted_nonzero <- aging_arima$predicted.nonzero

dim(Covfitted_nonzero)
dim(Covpredicted_nonzero)
```

```{r}
library(dplyr)
library(tidyverse)
clus.distance.arima = as.dist(1 - cor(t(Covpredicted_nonzero), use = "pa"))
h.clust.arima = hclust(clus.distance.arima,method='complete') # calc. clusters

cov_gene_cluster <- cutree(h.clust.arima, k = 4) %>% 
  # turn the named vector into a tibble
  enframe() %>% 
  # rename some of the columns
  dplyr::rename(gene = name, cluster = value)

head(cov_gene_cluster)
sum(cov_gene_cluster$cluster == 1)
sum(cov_gene_cluster$cluster == 2)
sum(cov_gene_cluster$cluster == 3)
sum(cov_gene_cluster$cluster == 4)

overlapgenes <- merge(predicted_nonzero, Covpredicted_nonzero, by="row.names")
dim(overlapgenes)
```

```{r}
Covfitted_nonzero <- as.data.frame(Covfitted_nonzero)
Covfitted_nonzero$cluster <- cov_gene_cluster$cluster
dim(Covfitted_nonzero)

cov_cluster1_names <- subset(cov_gene_cluster, cov_gene_cluster$cluster == 1)
cov_cluster1 <- Covfitted_nonzero[Covfitted_nonzero[,88]==1,c(1:87)]
rownames(cov_cluster1) <- cov_cluster1_names$gene
cov_cluster1 <- as.matrix(cov_cluster1)
# longcluster1 <- melt(cluster1)
# longcluster1 <- na.omit(longcluster1)

par(mar=c(1,1,1,1))
heatmap(cov_cluster1, Colv = NA, Rowv = NA)

cov_cluster2_names <- subset(cov_gene_cluster, cov_gene_cluster$cluster == 2)
cov_cluster2 <- Covfitted_nonzero[Covfitted_nonzero[,88]==2,c(1:87)]
rownames(cov_cluster2) <- cov_cluster2_names$gene
cov_cluster2 <- as.matrix(cov_cluster2)

heatmap(cov_cluster2, Colv = NA, Rowv = NA)

cov_cluster3_names <- subset(cov_gene_cluster, cov_gene_cluster$cluster == 3)
cov_cluster3 <- Covfitted_nonzero[Covfitted_nonzero[,88]==3,c(1:87)]
rownames(cov_cluster3) <- cov_cluster3_names$gene
cov_cluster3 <- as.matrix(cov_cluster3)

heatmap(cov_cluster3, Colv = NA, Rowv = NA)

cov_cluster4_names <- subset(cov_gene_cluster, cov_gene_cluster$cluster == 4)
cov_cluster4 <- Covfitted_nonzero[Covfitted_nonzero[,88]==4,c(1:87)]
rownames(cov_cluster4) <- cov_cluster4_names$gene
cov_cluster4 <- as.matrix(cov_cluster4)

heatmap(cov_cluster4, Colv = NA, Rowv = NA)

# write_csv(cov_cluster1_names, "CovCluster1.csv")
# write_csv(cov_cluster2_names, "CovCluster2.csv")
# write_csv(cov_cluster3_names, "CovCluster3.csv")
# write_csv(cov_cluster4_names, "CovCluster4.csv")
```

```{r}
#Making plots showing temporal patterns of gene expressionn

# cov_cluster3_fig <- melt(cov_cluster3)
# cov_cluster3_fig$Var2<-gsub("X","",as.character(cov_cluster3_fig$Var2))
# cov_cluster3_fig$Var2 <- as.numeric(as.character(cov_cluster3_fig$Var2))
# sapply(cov_cluster3_fig, class)
# cov_cluster3_fig <- melt(cov_cluster3, id="Var2")
# 
# ggplot(cov_cluster3_fig, aes(x=Var2, y=value)) +
#   geom_point() +
#   labs(x="Age", y = "Expression pattern")+
#   theme_minimal()
# 
# 
# cov_cluster4_fig <- melt(cov_cluster4)
# cov_cluster4_fig$Var2<-gsub("X","",as.character(cov_cluster4_fig$Var2))
# cov_cluster4_fig$Var2 <- as.numeric(as.character(cov_cluster4_fig$Var2))
# sapply(cov_cluster4_fig, class)
# cov_cluster4_fig <- melt(cov_cluster4, id="Var2")
# 
# ggplot(cov_cluster4_fig, aes(x=Var2, y=value)) +
#   geom_line() +
#   geom_point() + 
#   labs(x="Age", y = "Expression pattern")+
#   theme_minimal()


# Melt the data into a "long" format
cov1_fig <- cov_cluster1
colnames(cov1_fig) <- gsub("X", "", colnames(cov1_fig))
gene_expression_melted <- reshape2::melt(cov1_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression1 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

avg1 <- ggplot(average_expression1, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")
avg1

ggsave(filename = "CovCluster1_expression_plot.pdf", avg1)

cov2_fig <- cov_cluster2
colnames(cov2_fig) <- gsub("X", "", colnames(cov2_fig))
gene_expression_melted <- reshape2::melt(cov2_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression2 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

avg2 <- ggplot(average_expression2, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")
avg2

ggsave(filename = "CovCluster2_expression_plot.pdf", avg2)

cov3_fig <- cov_cluster3
colnames(cov3_fig) <- gsub("X", "", colnames(cov3_fig))
gene_expression_melted <- reshape2::melt(cov3_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression3 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

avg3 <- ggplot(average_expression3, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")
avg3

ggsave(filename = "CovCluster3_expression_plot.pdf", avg3)

cov4_fig <- cov_cluster4
colnames(cov4_fig) <- gsub("X", "", colnames(cov4_fig))
gene_expression_melted <- reshape2::melt(cov4_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression4 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

avg4 <- ggplot(average_expression4, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")
avg4

ggsave(filename = "CovCluster4_expression_plot.pdf", avg4)

list_df = list(average_expression1, average_expression2, average_expression3, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

cov_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")
plot(cov_average)

ggsave(filename = "Cov_expression_plot.pdf", cov_average)

list_df = list(average_expression3, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

cov_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(x = "Age",
       y = "Expression trajectory")
plot(cov_average)

ggsave(filename = "CovNonlinear_expression_plot.pdf", cov_average)

```

```{r}
cluster1overlap <- merge(cluster1, cov_cluster1, by="row.names")
dim(cluster1overlap)
cluster2overlap <- merge(cluster2, cov_cluster2, by="row.names")
dim(cluster2overlap)
cluster3overlap <- merge(cluster3, cov_cluster3, by="row.names")
dim(cluster3overlap)
cluster4overlap <- merge(cluster4, cov_cluster4, by="row.names")
dim(cluster4overlap)
```

```{r}
# Splitting male and female datasets and seeing if ARIMA models identify clusters 
# Making our input dataframes
female_metadata <- filter(covariates, sex == "F")
male_metadata <- filter(covariates, sex == "M")
head(male_metadata)

females <- female_metadata$sampid
males <- male_metadata$sampid

#female_relatedness <- relatedness[c(female_metadata$SubjectID),c(female_metadata$SubjectID)]
#z_female <- Z_mat[1:60,1:60]
#dge_female <- v %>% select(females)

#regress out effects of chips and then split count matrices
design <- model.matrix(~covariates$chip)
fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)

dge_female <- subset(resid_counts, select = c(female_metadata$sampid))

#male_relatedness <- relatedness[c(male_metadata$SubjectID),c(male_metadata$SubjectID)]
#z_male <- Z_mat[1:36,1:36]
dge_male <- subset(resid_counts, select = c(male_metadata$sampid))


save(dge_male, male_metadata, file = "MaleInputDataARIMA_2.RData")
save(dge_female, female_metadata, file = "FemaleInputDataARIMA_2.RData")

```

```{r}
all_genes <- as.data.frame(rownames(v))
write_csv(all_genes, "AllGenesList.csv")
```

```{r}
load("ARIMA_2/maleaging_arima.RData")

malefitted_nonzero <- aging_arima$fitted.nonzero
malepredicted_nonzero <- aging_arima$predicted.nonzero

dim(malefitted_nonzero)
dim(malepredicted_nonzero)

overlapgenes <- merge(malepredicted_nonzero, Covpredicted_nonzero, by="row.names")
dim(overlapgenes)
```

```{r}
load("ARIMA_2/femaleaging_arima.RData")

femalefitted_nonzero <- aging_arima$fitted.nonzero
femalepredicted_nonzero <- aging_arima$predicted.nonzero

dim(femalefitted_nonzero)
dim(femalepredicted_nonzero)

overlapgenes <- merge(femalepredicted_nonzero, Covpredicted_nonzero, by="row.names")
dim(overlapgenes)
```

```{r}
library(heatmap3)
clus.distance.arima = as.dist(1 - cor(t(femalepredicted_nonzero), use = "pa"))
h.clust.arima = hclust(clus.distance.arima,method='complete') # calc. clusters

female_gene_cluster <- cutree(h.clust.arima, k = 4) %>% 
  # turn the named vector into a tibble
  enframe() %>% 
  # rename some of the columns
  rename(gene = name, cluster = value)

head(female_gene_cluster)

```

```{r}
femalefitted_nonzero <- as.data.frame(femalefitted_nonzero)
femalefitted_nonzero$cluster <- female_gene_cluster$cluster
dim(femalefitted_nonzero)

female_cluster1_names <- subset(female_gene_cluster, female_gene_cluster$cluster == 1)
female_cluster1 <- femalefitted_nonzero[femalefitted_nonzero[,56]==1,c(1:55)]
rownames(female_cluster1) <- female_cluster1_names$gene
female_cluster1 <- as.matrix(female_cluster1)
# longcluster1 <- melt(cluster1)
# longcluster1 <- na.omit(longcluster1)

par(mar=c(1,1,1,1))
heatmap(female_cluster1, Colv = NA, Rowv = NA)

female_cluster2_names <- subset(female_gene_cluster, female_gene_cluster$cluster == 2)
female_cluster2 <- femalefitted_nonzero[femalefitted_nonzero[,56]==2,c(1:55)]
rownames(female_cluster2) <- female_cluster2_names$gene
female_cluster2 <- as.matrix(female_cluster2)

heatmap(female_cluster2, Colv = NA, Rowv = NA)

female_cluster3_names <- subset(female_gene_cluster, female_gene_cluster$cluster == 3)
female_cluster3 <- femalefitted_nonzero[femalefitted_nonzero[,56]==3,c(1:55)]
rownames(female_cluster3) <- female_cluster3_names$gene
female_cluster3 <- as.matrix(female_cluster3)

heatmap(female_cluster3, Colv = NA, Rowv = NA)

female_cluster4_names <- subset(female_gene_cluster, female_gene_cluster$cluster == 4)
female_cluster4 <- femalefitted_nonzero[femalefitted_nonzero[,56]==4,c(1:55)]
rownames(female_cluster4) <- female_cluster4_names$gene
female_cluster4 <- as.matrix(female_cluster4)

heatmap(female_cluster4, Colv = NA, Rowv = NA)

write_csv(female_cluster1_names, "female_cluster1.csv")
write_csv(female_cluster2_names, "female_cluster2.csv")
write_csv(female_cluster3_names, "female_cluster3.csv")
write_csv(female_cluster4_names, "female_cluster4.csv")

```

```{r}
#Making expression figs for female data
# Melt the data into a "long" format
female1_fig <- female_cluster1
colnames(female1_fig) <- gsub("X", "", colnames(female1_fig))
gene_expression_melted <- reshape2::melt(female1_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression1 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression1, aes(x = Var2, y = Expression)) +
  geom_point(color = "#9a5ea1", show.legend = FALSE) + # Setting point color
  geom_line(color = "#9a5ea1", show.legend = FALSE) + # Setting line color
  theme_minimal() +
  ylim(-0.3, 0.23) +
  labs(x = "Age",
       y = "Expression trajectory")


female2_fig <- female_cluster2
colnames(female2_fig) <- gsub("X", "", colnames(female2_fig))
gene_expression_melted <- reshape2::melt(female2_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression2 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression2, aes(x = Var2, y = Expression)) +
  geom_point(color = "#9a5ea1", show.legend = FALSE) + # Setting point color
  geom_line(color = "#9a5ea1", show.legend = FALSE) + # Setting line color
  theme_minimal() +
  ylim(-0.3, 0.23) +
  labs(x = "Age",
       y = "Expression trajectory")

female3_fig <- female_cluster3
colnames(female3_fig) <- gsub("X", "", colnames(female3_fig))
gene_expression_melted <- reshape2::melt(female3_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression3 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression3, aes(x = Var2, y = Expression)) +
  geom_point(color = "#9a5ea1", show.legend = FALSE) + # Setting point color
  geom_line(color = "#9a5ea1", show.legend = FALSE) + # Setting line color
  theme_minimal() +
  ylim(-0.3, 0.23) +
  labs(x = "Age",
       y = "Expression trajectory")

female4_fig <- female_cluster4
colnames(female4_fig) <- gsub("X", "", colnames(female4_fig))
gene_expression_melted <- reshape2::melt(female4_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression4 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression4, aes(x = Var2, y = Expression)) +
  geom_point(color = "#9a5ea1", show.legend = FALSE) + # Setting point color
  geom_line(color = "#9a5ea1", show.legend = FALSE) + # Setting line color
  theme_minimal() +
  ylim(-0.3, 0.23) +
  labs(x = "Age",
       y = "Expression trajectory")

list_df = list(average_expression1, average_expression2, average_expression3, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

female_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")
plot(female_average)

ggsave(filename = "Female_expression_plot.pdf", female_average)

list_df = list(average_expression1, average_expression3)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

female_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  scale_color_manual(values=c("#8c96c6", "#88419d")) +
  theme_light() +
  ylim(-0.15, 0.175) +
  labs(x = "Age",
       y = "Expression trajectory")
plot(female_average)

ggsave(filename = "FemaleNonlinear_expression_plot.pdf", female_average)
```

```{r}
clus.distance.arima = as.dist(1 - cor(t(malepredicted_nonzero), use = "pa"))
h.clust.arima = hclust(clus.distance.arima,method='complete') # calc. clusters

male_gene_cluster <- cutree(h.clust.arima, k = 4) %>% 
  # turn the named vector into a tibble
  enframe() %>% 
  # rename some of the columns
  rename(gene = name, cluster = value)

head(male_gene_cluster)

malefitted_nonzero <- as.data.frame(malefitted_nonzero)
malefitted_nonzero$cluster <- male_gene_cluster$cluster
dim(malefitted_nonzero)

male_cluster1_names <- subset(male_gene_cluster, male_gene_cluster$cluster == 1)
male_cluster1 <- malefitted_nonzero[malefitted_nonzero[,33]==1,c(1:32)]
rownames(male_cluster1) <- male_cluster1_names$gene
male_cluster1 <- as.matrix(male_cluster1)

par(mar=c(1,1,1,1))
heatmap(male_cluster1, Colv = NA, Rowv = NA)

male_cluster2_names <- subset(male_gene_cluster, male_gene_cluster$cluster == 2)
male_cluster2 <- malefitted_nonzero[malefitted_nonzero[,33]==2,c(1:32)]
rownames(male_cluster2) <- male_cluster2_names$gene
male_cluster2 <- as.matrix(male_cluster2)

heatmap(male_cluster2, Colv = NA, Rowv = NA)

male_cluster3_names <- subset(male_gene_cluster, male_gene_cluster$cluster == 3)
male_cluster3 <- malefitted_nonzero[malefitted_nonzero[,33]==3,c(1:32)]
rownames(male_cluster3) <- male_cluster3_names$gene
male_cluster3 <- as.matrix(male_cluster3)

heatmap(male_cluster3, Colv = NA, Rowv = NA)

male_cluster4_names <- subset(male_gene_cluster, male_gene_cluster$cluster == 4)
male_cluster4 <- malefitted_nonzero[malefitted_nonzero[,33]==4,c(1:32)]
rownames(male_cluster4) <- male_cluster4_names$gene
male_cluster4 <- as.matrix(male_cluster4)

heatmap(male_cluster4, Colv = NA, Rowv = NA)

write_csv(male_cluster1_names, "male_cluster1.csv")
write_csv(male_cluster2_names, "male_cluster2.csv")
write_csv(male_cluster3_names, "male_cluster3.csv")
write_csv(male_cluster4_names, "male_cluster4.csv")
```

```{r}
#Making expression figs
# Melt the data into a "long" format
male1_fig <- male_cluster1
colnames(male1_fig) <- gsub("X", "", colnames(male1_fig))
gene_expression_melted <- reshape2::melt(male1_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression1 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression1, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE, color = "#98823c") +  # Change point color
  geom_line(show.legend = FALSE, color = "#98823c") +   # Change line color
  theme_minimal() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")

male2_fig <- male_cluster2
colnames(male2_fig) <- gsub("X", "", colnames(male2_fig))
gene_expression_melted <- reshape2::melt(male2_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression2 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression2, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE, color = "#98823c") +  # Change point color
  geom_line(show.legend = FALSE, color = "#98823c") +   # Change line color
  theme_minimal() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")

male3_fig <- male_cluster3
colnames(male3_fig) <- gsub("X", "", colnames(male3_fig))
gene_expression_melted <- reshape2::melt(male3_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression3 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression3, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE, color = "#98823c") +  # Change point color
  geom_line(show.legend = FALSE, color = "#98823c") +   # Change line color
  theme_minimal() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")

male4_fig <- male_cluster4
colnames(male4_fig) <- gsub("X", "", colnames(male4_fig))
gene_expression_melted <- reshape2::melt(male4_fig, id.vars = "Gene", variable.name = "Age", value.name = "Expression")
average_expression4 <- aggregate(Expression ~ Var2, gene_expression_melted, mean)

ggplot(gene_expression_melted, aes(x = Var2, y = Expression, group = Var1, color = Var1)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")

ggplot(average_expression4, aes(x = Var2, y = Expression)) +
  geom_point(show.legend = FALSE, color = "#98823c") +  # Change point color
  geom_line(show.legend = FALSE, color = "#98823c") +   # Change line color
  theme_minimal() +
  ylim(-0.3, 0.2) +
  labs(x = "Age",
       y = "Expression trajectory")

list_df = list(average_expression1, average_expression2, average_expression3, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

male_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_minimal() + 
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")
plot(male_average)

ggsave(filename = "Male_expression_plot.pdf", male_average)

list_df = list(average_expression1, average_expression2, average_expression3, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

male_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  labs(title = "Gene Expression over Time",
       x = "Age",
       y = "Expression trajectory")
plot(male_average)

ggsave(filename = "Male_expression_plot.pdf", male_average)

list_df = list(average_expression1, average_expression4)
average_combined <- list_df %>% reduce(inner_join, by='Var2')
average_melted <- reshape2::melt(average_combined, id.vars = "Var2", variable.name = "Cluster", value.name = "Expression")

male_average <- ggplot(average_melted, aes(x = Var2, y = Expression, group = Cluster, color = Cluster)) +
  geom_point(show.legend = FALSE) +
  geom_line(show.legend = FALSE) +
  theme_light() +
  scale_color_manual(values=c("#41b6c4", "#225ea8")) +
  ylim(-0.15, 0.175) + 
  labs(x = "Age",
       y = "Expression trajectory")
plot(male_average)

ggsave(filename = "MaleNonlinear_expression_plot.pdf", male_average)
```

```{r}
malefemaleoverlap <- merge(femalefitted_nonzero, malefitted_nonzero, by="row.names")
dim(malefemaleoverlap)
female_covoverlap <- merge(femalefitted_nonzero, Covfitted_nonzero, by="row.names")
dim(female_covoverlap)
male_covoverlap <- merge(malefitted_nonzero, Covfitted_nonzero, by="row.names")
dim(male_covoverlap)

cluster1overlap <- merge(male_cluster2, femalefitted_nonzero, by="row.names")
dim(cluster1overlap)
cluster2overlap <- merge(male_cluster1, femalefitted_nonzero, by="row.names")
dim(cluster2overlap)
cluster3overlap <- merge(male_cluster3, female_cluster4, by="row.names")
dim(cluster3overlap)
cluster4overlap <- merge(male_cluster4, femalefitted_nonzero, by="row.names")
dim(cluster4overlap)

cluster3overlap <- merge(male_cluster3, femalefitted_nonzero, by="row.names")
dim(cluster3overlap)
cluster3overlap <- merge(malefitted_nonzero, femalefitted_nonzero, by="row.names")
dim(cluster3overlap)
```

```{r}
cluster1overlap <- merge(male_cluster1, Covfitted_nonzero, by="row.names")
dim(cluster1overlap)
cluster2overlap <- merge(male_cluster2, Covfitted_nonzero, by="row.names")
dim(cluster2overlap)
cluster3overlap <- merge(female_cluster1, Covfitted_nonzero, by="row.names")
dim(cluster3overlap)
cluster4overlap <- merge(female_cluster2, Covfitted_nonzero, by="row.names")
dim(cluster4overlap)


cluster1overlap <- merge(malefitted_nonzero, cov_cluster1, by="row.names")
dim(cluster1overlap)
cluster2overlap <- merge(malefitted_nonzero, cov_cluster2, by="row.names")
dim(cluster2overlap)
cluster3overlap <- merge(femalefitted_nonzero, cov_cluster1, by="row.names")
dim(cluster3overlap)
cluster4overlap <- merge(femalefitted_nonzero, cov_cluster2, by="row.names")
dim(cluster4overlap)

```

```{r}
female_covoverlap <- merge(femalefitted_nonzero, Covfitted_nonzero, by="row.names")
dim(female_covoverlap)
male_covoverlap <- merge(malefitted_nonzero, Covfitted_nonzero, by="row.names")
dim(male_covoverlap)

male_genes <- as.data.frame(rownames(malefitted_nonzero))
colnames(male_genes) <- c('gene')
male_overlap_genes <- as.data.frame((male_covoverlap$Row.names))
colnames(male_overlap_genes) <- c('gene')
malefemaleoverlap <- merge(femalefitted_nonzero, malefitted_nonzero, by="row.names")
malefemaleoverlap_genes <- as.data.frame(malefemaleoverlap$Row.names)
colnames(malefemaleoverlap_genes) <- c('gene')

male_unique <- anti_join(male_genes, male_overlap_genes, by='gene')
male_unique <- anti_join(male_unique, malefemaleoverlap_genes, by='gene')
dim(male_unique)

female_genes <- as.data.frame(rownames(femalefitted_nonzero))
colnames(female_genes) <- c('gene')
female_overlap_genes <- as.data.frame((female_covoverlap$Row.names))
colnames(female_overlap_genes) <- c('gene')

female_unique <- anti_join(female_genes, female_overlap_genes, by='gene')
female_unique <- anti_join(female_unique, malefemaleoverlap_genes, by='gene')
dim(female_unique)

unique_overlap <- merge(female_unique, male_unique, by='gene')

rownames(male_unique) <- male_unique$gene
rownames(female_unique) <- female_unique$gene

male_unique_overlap <- merge(male_unique, male_cluster1, by='row.names')
dim(male_unique_overlap)
male_unique_overlap <- merge(male_unique, male_cluster2, by='row.names')
dim(male_unique_overlap)
male_unique_overlap <- merge(male_unique, male_cluster3, by='row.names')
dim(male_unique_overlap)
male_unique_overlap <- merge(male_unique, male_cluster4, by='row.names')
dim(male_unique_overlap)

female_unique_overlap <- merge(female_unique, female_cluster1, by='row.names')
dim(female_unique_overlap)
female_unique_overlap <- merge(female_unique, female_cluster2, by='row.names')
dim(female_unique_overlap)
female_unique_overlap <- merge(female_unique, female_cluster3, by='row.names')
dim(female_unique_overlap)
female_unique_overlap <- merge(female_unique, female_cluster4, by='row.names')
dim(female_unique_overlap)


write_csv(male_unique, file='male_unique_genes.csv')
write_csv(female_unique, file='female_unique_genes.csv')
```

```{r}
#rerunning the analysis to test statistical power for a subset (n=40) of individuals 

# Import necessary libraries
library(dplyr)

# Assume the dataset is called "data"

# Find the range of ages in the dataset
age_range <- range(covariates$age)

# Sample 40 individuals randomly while representing the entire age range
sampled_data <- covariates %>%
  filter(age >= age_range[1] & age <= age_range[2]) %>%
  sample_n(60)

#regress out the effects of chips and sex from dataset and then filter to only inlcude the 40 individuals in our sample dataset
design <- model.matrix(~covariates$chip + covariates$sex)
fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)

dge_Test <- subset(resid_counts, select = c(sampled_data$sampid))

save(dge_Test, sampled_data, file = "TESTInputDataARIMA_3.RData")
```

```{r}
#Next we can analyze the results of the test function to see if our results hold with a smaller sample size 

load("ARIMA_2/TEST_arima_3.RData")

TESTfitted_nonzero <- aging_arima$fitted.nonzero
TESTpredicted_nonzero <- aging_arima$predicted.nonzero

dim(TESTfitted_nonzero)
dim(TESTpredicted_nonzero)

library(heatmap3)
clus.distance.arima = as.dist(1 - cor(t(TESTpredicted_nonzero), use = "pa"))
h.clust.arima = hclust(clus.distance.arima,method='complete') # calc. clusters

library(dplyr)
library(tidyverse)

TEST_gene_cluster <- cutree(h.clust.arima, k = 4) %>% 
  # turn the named vector into a tibble
  enframe() %>% 
  # rename some of the columns
  rename(gene = name, cluster = value)

head(TEST_gene_cluster)

TESTfitted_nonzero <- as.data.frame(TESTfitted_nonzero)
TESTfitted_nonzero$cluster <- TEST_gene_cluster$cluster
dim(TESTfitted_nonzero)

TEST_cluster1_names <- subset(TEST_gene_cluster, TEST_gene_cluster$cluster == 1)
TEST_cluster1 <- TESTfitted_nonzero[TESTfitted_nonzero[,61]==1,c(1:60)]
rownames(TEST_cluster1) <- TEST_cluster1_names$gene
TEST_cluster1 <- as.matrix(TEST_cluster1)

par(mar=c(1,1,1,1))
heatmap(TEST_cluster1, Colv = NA, Rowv = NA)

TEST_cluster2_names <- subset(TEST_gene_cluster, TEST_gene_cluster$cluster == 2)
TEST_cluster2 <- TESTfitted_nonzero[TESTfitted_nonzero[,61]==2,c(1:60)]
rownames(TEST_cluster2) <- TEST_cluster2_names$gene
TEST_cluster2 <- as.matrix(TEST_cluster2)

heatmap(TEST_cluster2, Colv = NA, Rowv = NA)

TEST_cluster3_names <- subset(TEST_gene_cluster, TEST_gene_cluster$cluster == 3)
TEST_cluster3 <- TESTfitted_nonzero[TESTfitted_nonzero[,61]==3,c(1:60)]
rownames(TEST_cluster3) <- TEST_cluster3_names$gene
TEST_cluster3 <- as.matrix(TEST_cluster3)

heatmap(TEST_cluster3, Colv = NA, Rowv = NA)

TEST_cluster4_names <- subset(TEST_gene_cluster, TEST_gene_cluster$cluster == 4)
TEST_cluster4 <- TESTfitted_nonzero[TESTfitted_nonzero[,61]==4,c(1:60)]
rownames(TEST_cluster4) <- TEST_cluster4_names$gene
TEST_cluster4 <- as.matrix(TEST_cluster4)

heatmap(TEST_cluster4, Colv = NA, Rowv = NA)

TEST_covoverlap <- merge(TESTfitted_nonzero, Covfitted_nonzero, by="row.names")
```


Rerunning ARIMAs models with regressin out effects of cell type proportions
```{r}
cell_type <- read_csv(file = "Cibersort/CIBERSORTx_Job13_Results_updated_FINALFORPAPER.csv")
cell_type$sex <- covariates$sex
cell_type$chip <- covariates$chip
colnames(cell_type)[colnames(cell_type) == "Individual"] <- "sampid"
colnames(cell_type)[colnames(cell_type) == "Age"] <- "age"

design <- model.matrix(~cell_type$chip + cell_type$sex + cell_type$astrocytes + cell_type$`basket cells` + cell_type$`GABAergic neurons` + cell_type$`glutamatergic neurons` + cell_type$`medium spiny neurons` + cell_type$microglia + cell_type$oligodendrocytes + cell_type$`vascular cells`)

#design <- model.matrix(~cell_type$chip + cell_type$sex + cell_type$`glutamatergic neurons` + cell_type$oligodendrocytes)

fit <-lmFit(v,design)
fit <- eBayes(fit)
resid_counts <-residuals.MArrayLM(object=fit, v)
resid_counts <-resid_counts[order(rownames(resid_counts)),]
resid_counts <- na.omit(resid_counts)

save(resid_counts, cell_type, file = "InputDataARIMA_Cibersort.RData")
```